[asl] Binary operator for string concatenation

hrutvik opened the pull request:
Support string concatenation by overloading the existing `::` binary operator (previously only bit vector concatenation).

Roman-Manevich commented on asllib/Operations.ml:
```diff
@@ -108,6 +108,10 @@ let binop_values pos t (op : binop) v1 v2 =
   | `EQ_OP, L_Label s1, L_Label s2 -> L_Bool (String.equal s1 s2)
   | `NEQ, L_Label s1, L_Label s2 -> L_Bool (not (String.equal s1 s2))
+  (* string concatenation *)
+  | `CONCAT, _, _ ->
```
Why not use `L_String` instead of `_` for both operands?

Roman-Manevich commented on asllib/Typing.ml:
```diff
@@ -658,8 +658,18 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
+    | `CONCAT, (T_Bits (w1, _), T_Bits (w2, _)) ->
         T_Bits (width_plus env w1 w2, []) |> here
+    | `CONCAT, _ ->
+        let+ () =
```
Why the check for singular types rather than the more specific check that each operand type is string?

Roman-Manevich commented on asllib/doc/PrimitiveOperations.tex:
```diff
@@ -776,7 +776,8 @@ \section{Typing\label{sec:PrimitiveOperationsTyping}}
   (\PLUS      &,& \lbitvector &,& \lbitvector)&,\\
-  (\BVCONCAT  &,& \lbitvector &,& \lbitvector)&,\\
+  (\CONCAT    &,& \lbitvector &,& \lbitvector)&,\\
+  (\CONCAT    &,& \Ignore &,& \Ignore)&,\\
```
I think the operand columns should be `\string`.

Roman-Manevich commented on asllib/doc/PrimitiveOperations.tex:
```diff
@@ -1142,6 +1143,15 @@ \section{Typing\label{sec:PrimitiveOperationsTyping}}
 
+  \item \AllApplyCase{concat\_strings}
+  \begin{itemize}
+    \item $\vvone$ and $\vvtwo$ are not both bitvector literals;
```
please use a positive check that both those are string literals.

Roman-Manevich commented on asllib/doc/PrimitiveOperations.tex:
```diff
@@ -1539,6 +1549,21 @@ \subsubsection{Operators Over String Values \label{sec:StringOperations}}
 
+\begin{mathpar}
+\inferrule[concat\_strings]{
+  \vvone \neq \lbitvector(\Ignore) \lor \vvtwo \neq \lbitvector(\Ignore) \\
```
same here (both operands are string literals).

Roman-Manevich commented on asllib/doc/RelationsOnTypes.tex:
```diff
@@ -1119,14 +1119,23 @@ \section{Relations Over Types\label{sec:RelationsOnTypes}}
+  \item \AllApplyCase{string\_concat}
+  \begin{itemize}
+    \item $\op$ is $\CONCAT$;
+    \item $\vtone$ and $\vttwo$ are not both bit vector types;
```
same here (both operands are string types).

Roman-Manevich commented on asllib/doc/RelationsOnTypes.tex:
```diff
@@ -1236,7 +1245,7 @@ \section{Relations Over Types\label{sec:RelationsOnTypes}}
       \PLUS    & \TBits  & \TBits\\
       \MINUS   & \TBits  & \TBits\\
-      \BVCONCAT & \TBits  & \TBits\\
+      \CONCAT  & \Ignore & \Ignore\\
```
same here `\TString`.

Roman-Manevich commented on asllib/doc/RelationsOnTypes.tex:
```diff
@@ -1327,11 +1336,22 @@ \section{Relations Over Types\label{sec:RelationsOnTypes}}
 
+\begin{mathpar}
+\inferrule[string\_concat]{
+  \vtone \neq \TBits(\Ignore, \Ignore) \lor \vttwo \neq \TBits(\Ignore, \Ignore) \\
```
same here - positive check for string types.

Roman-Manevich commented on asllib/doc/RelationsOnTypes.tex:
```diff
@@ -1452,7 +1472,7 @@ \section{Relations Over Types\label{sec:RelationsOnTypes}}
       (\PLUS    &,& \TBits  &,& \TBits)\\
       (\MINUS   &,& \TBits  &,& \TBits)\\
-      (\BVCONCAT &,& \TBits  &,& \TBits)\\
+      (\CONCAT  &,& \TBits  &,& \TBits)\\
```
add a row for `\CONCAT` of string types.

Roman-Manevich reviewed (commented):
- Please add string concatenation examples to `TypingRule.ApplyBinopTypes.asl` and
`TypingRule.BinopLiterals.string_labels.asl`
- Please add an entry in the ALP3.1 changelog.

hrutvik commented:
@HadrienRenaud I have reverted the collapsing of the printing functions. I have moved the one for printing to the command line to `Operations.ml` instead.

@Roman-Manevich in response to your comments:
- Regarding the checks on types, the concatenation operator is intended to work for all singular types, by first converting them to strings (in the same way that print statements convert them to strings). Please do let me know how this affects your review comments.
- I have added examples in the files you suggest.
- I will follow up with a separate PR for the changelog if this is OK, consolidating the changes of the various open PRs I currently have (to avoid the possibility of conflicts).

Roman-Manevich commented:
@hrutvik I wasn't aware of the automatic conversion of singular types to strings. I now think that the tables and rules make sense, but I also think this conversion should be explained in the reference and probably deserves a specific example.

hrutvik commented:
I've added an example explaining the conversion and pointing to the listing!
